agent/src/capability/baseconfigstore.cpp

```diff
@@ -23,173 +23,201 @@ namespace softwarecontainer {
 
 BaseConfigStore::BaseConfigStore(const std::string &inputPath)
 {
-    ReturnCode retval = ReturnCode::SUCCESS;
-
     if (inputPath.empty()) {
         //File path is empty, this is ok
         return;
     }
 
     if (isDirectory(inputPath)) {
-        retval = readCapsFromDir(inputPath);
+        try {
+            readCapsFromDir(inputPath);
+        } catch(ConfigStoreError &error) {
+            throw;
+        }
     } else if (isFile(inputPath)) {
-        retval = readCapsFromFile(inputPath);
+        try {
+            readCapsFromFile(inputPath);
+        } catch(ConfigStoreError &error) {
+            throw;
+        }
     } else {
         //Could not find a matching clause
         log_error() << "\"" << inputPath << "\"" << " does not exist";
-        retval = ReturnCode::FAILURE;
-    }
-
-    if (isError(retval)) {
-        throw ReturnCode::FAILURE;
+        throw ConfigStoreError("Path to service manifest does not exist");
     }
 }
 
 BaseConfigStore::~BaseConfigStore()
 {
 }
 
-ReturnCode BaseConfigStore::readCapsFromDir(const std::string &dirPath)
+void BaseConfigStore::readCapsFromDir(const std::string &dirPath)
 {
+    std::string errorMessage;
+
     if (dirPath.compare("/") == 0) {
-        log_warning() << "Searching for configuration files from root dir not allowed";
-        return ReturnCode::FAILURE;
+        errorMessage = "Searching for configuration files from root dir not allowed";
+        log_warning() << errorMessage;
+        throw ServiceManifestPathError(errorMessage);
     }
     if (isDirectoryEmpty(dirPath)) {
         log_warning() << "Path to configuration files is empty: " << dirPath;
-        return ReturnCode::SUCCESS;
+        return;
     }
 
     std::vector<std::string> files = fileList(dirPath);
 
     if (files.empty()) {
         log_info() << "No configuration files found: " << dirPath;
-        return ReturnCode::SUCCESS;
+        return;
     }
 
     for (std::string file : files) {
         std::string filePath = dirPath + file;
 
-        if (isError(readCapsFromFile(filePath))) {
-            log_warning() << "Could not parse a file in directory: " << filePath;
-            return ReturnCode::FAILURE;
+        try {
+            readCapsFromFile(filePath);
+        } catch(ConfigStoreError &error) {
+            errorMessage = "Could not parse a file in directory";
+            log_warning() << errorMessage << ": " << filePath;
+            throw ServiceManifestPathError(errorMessage);
```

Comment:
I don't really understand the thinking behind the error types. It should be commented or documented what kinds of errors happen when and what they mean. For example readCapsFromFile will throw any of the two exceptions it can throw and in some code paths it is just propagated, and in some it is caught and made into a specific exception.